### PR TITLE
Implements seekDir and getNextFileName on FS Lib to improve performance

### PR DESCRIPTION
**Pull request by Lucasczm**
## Description of Change
Implemented two functions: 
 - getNextFileName:  Get the next file name with better performance than getNextFile function. To improve file listing.
 - seekDir: Set the position of directory for improve search.

Examples: 
```cpp
 File root = SD.open("/data");
 String filename;
 //Return true if have next file, false if not
 if (root.getNextFileName(filename)) {
     Serial.println(filename); //print "file1"
 }
```

## Tests scenarios
Tested on PlatformIO and Arduino core v2.0.4 with  ESP32 LILYGO PoE Board

## Related links

#5002

### CODE REVIEW

**Comment by me-no-dev** [comment]
For `getNextFileName` why don't you return the filename directly (or empty String if not available) instead?

File: libraries/FS/src/FS.cpp
Comment (SuGlider) [inline libraries/FS/src/FS.cpp:200]: I see an inconsistency here given that the function `getNextFileName()` shall return a `boolean`.
it is better to `return false` instead.

**Review by SuGlider — CHANGES_REQUESTED** [review]

File: libraries/FS/src/FS.cpp
Comment (me-no-dev) [reply libraries/FS/src/FS.cpp:200]: @SuGlider see my comment above. It's better to return `String` directly, not false or NULL.

Comment (SuGlider) [reply libraries/FS/src/FS.cpp:200]: Ok, it make the API easier returning a String as you proposed.

**Comment by PilnyTomas** [comment]
Tested using [this sketch](https://github.com/PilnyTomas/issues/blob/main/PR_7729/PR_7729.ino) confirming that the new function is way much faster:
new function ~0.24ms per folder read
old function `openNextFile` ~142ms per folder read

**Review by PilnyTomas — APPROVED** [review]
Apart from the requested changes looks good.

File: libraries/FS/src/FS.cpp
Comment (Lucasczm) [reply libraries/FS/src/FS.cpp]: i maked the changes, now its return `String` name directly or String empty

**Review by Lucasczm — COMMENTED** [review]
Now its return String name directly or String empty

**Comment by tueddy** [comment]
The speed increase is amazing! My webserver based SD fileexplorer needs 6.8 seconds with getNextFile(), now it's just 100ms.

But how can we decide wether it's a file or a (sub-)directory? My explorer JavaScript needs the information "isDir=false/true" 
I would expect the name has a trailing slash when it's a directory.

Or can it enhanced with a similar function getNextPathName() ?

**Comment by tueddy** [comment]
The speed improvement is amazing! For a webserver SD filexplorer the time for listing is currently about 2 seconds (very blocking), with this PR only 20 msecs (hundreds of mp3 files on SD card). Works fine here!

But for the Javascript based file explorer tree i also need the information wether it's a file or directory. Is there a way to get it?
Suggestion: add a trailing slash in filename if it's a directory or add an overload function with out-param isDirectory.
Some other information like size, creation date,  last mod-date would also be helpful for sorting the tree. 

Thank's!

**Comment by tueddy** [comment]
I made a speedtest example for PIO here https://github.com/tueddy/FS/tree/main/example
Tested my SD-cards with a lot of files in root directory:


Samsung Evo 32GB:
getNextFileName(), done reading root-directory 36 elemts in 18 ms = 0.50 ms/folder
openNextFile(), done reading root-directory 36 elements in 492 ms = 13.67 ms/elements

Kingston 8GB, a bit older card:
getNextFileName(), done reading root-directory 42 elemts in 29 ms = 0.69 ms/folder
openNextFile(), done reading root-directory 42 elements in 1119 ms = 26.64 ms/elements

Transcend 16GB:
getNextFileName(), done reading root-directory 84 elemts in 83 ms = 0.99 ms/folder
openNextFile(), done reading root-directory 84 elements in 6960 ms = 82.86 ms/elements

Scandisk 16GB:
getNextFileName(), done reading root-directory 151 elemts in 68 ms = 0.45 ms/folder
openNextFile(), done reading root-directory 151 elements in 6872 ms = 45.51 ms/elements

**Review by me-no-dev — APPROVED** [review]

**Comment by me-no-dev** [comment]
@SuGlider waiting on your approval here :)

**Review by P-R-O-C-H-Y — APPROVED** [review]

File: libraries/FS/src/vfs_api.cpp
Comment (vortigont) [inline libraries/FS/src/vfs_api.cpp:506]: just a small hint, this could be reduced to avoid a couple of extra copy's

```c++
    String fname(file->d_name);
    if (!fname.startsWith("/") && !name.endsWith("/")) {
        name += "/";
    }
    return String(_path)+fname;
```

**Review by vortigont — COMMENTED** [review]
a small hint for String's

File: libraries/FS/src/vfs_api.cpp
Comment (mrengineer7777) [reply libraries/FS/src/vfs_api.cpp:506]: @vortigont Your change loses `String name` declaration, and then doesn't use name in return. BUG.

Comment (vortigont) [reply libraries/FS/src/vfs_api.cpp:506]: right, it's the `name` in second condition. my bad, I should have my glasses on :) tnx!

**Review by SuGlider — APPROVED** [review]
@Lucasczm - Thanks for the PR! Very good.

**Comment by SuGlider** [comment]
@Lucasczm - I can't rebase it from here. Please do it in your end in order to be able to merge it. Thanks.

**Comment by tueddy** [comment]
Can't test it just at the moment, but can I distinguish whether in the listing is a file or directory with this commit?
Made a comment [above](https://github.com/espressif/arduino-esp32/pull/7229#issuecomment-1257071883)..

Thank's for your work!

**Comment by BlueAndi** [comment]
@tueddy The seekDir() implementation will only be useful if the other functions are available in FS and VFS too, see
https://github.com/espressif/esp-idf/blob/16a4ee7c36a848ca155791677ce011f3ca75c519/components/newlib/platform_include/sys/dirent.h#L51-L57

**Comment by tueddy** [comment]
getNextFileName() works fine in 2.0.6, but how can i detect now if return value is a file or directory?

File: libraries/FS/src/vfs_api.cpp
Comment (nomis) [inline libraries/FS/src/vfs_api.cpp:499]: This behaviour is inconsistent with `openNextFile()` which skips to the next file.

It's not possible to tell the difference between "end of file list" and "something that is not a file or directory".

File: libraries/FS/src/vfs_api.cpp
Comment (nomis) [inline libraries/FS/src/vfs_api.cpp:505]: This is the "path", not the "name" of the file.

This results in more effort by users of this function to remove the directory name (which they already have) from every filename.

(The `openNextFile()` function returns a `File` so it's possible to use `name()` or `path()`)

**Pull request merged**
